tests/cpp/test_multidevice_communications.cpp

```diff
@@ -58,121 +58,96 @@ void CommunicationTest::validate(at::Tensor obtained, at::Tensor expected) {
       obtained);
 }
 
-void CommunicationTest::resetDstBuffers() {
-  for (auto& buf : params.dst_bufs) {
-    buf.copy_(at::full(tensor_size, nan(""), tensor_options));
-  }
-}
-
 TEST_P(CommunicationTest, Gather) {
   params.root = root;
   params.team = all_ranks;
-  params.src_bufs = {at::empty(tensor_size, tensor_options)};
-  if (communicator->deviceId() == root) {
-    for (int64_t i = 0; i < communicator->size(); i++) {
-      params.dst_bufs.push_back(at::empty(tensor_size, tensor_options));
-    }
-  }
   auto communication = Gather(params);
 
-  for (int j : c10::irange(number_of_repetitions)) {
```

Comment:
I would personally revert this change
The implementation was made so that we keep the same src/dst buffers when repeating the collectives. This has the following advantages:
- We test/benefit  backend's buffer registration cache. This is essential for benchmarks
- We save allocation time when testing
- The tests are "harden" this wew really test that we can have many inflights collectives on the same buffers. This emulates more accurately the type of Workload we are targeting

Reply:
Makes sense!

Reply:
> Done -- the tests now reuse input/output buffers.

Thx!


> I'd love to learn more about that.

Buffer registration can come into play in several places. For example, in low level [RDMA infiniband verbs](https://www.rdmamojo.com/2012/09/07/ibv_reg_mr/), the buffers needs to be registered to the NIC before rdma can be posted. E.g., UCX (p2p library used by ucc and possibly nccl) uses such [registration cache](https://github.com/openucx/ucx/blob/e11197ce7e6d0d892a9f96300c789fa036aad6c8/docs/source/glossaries.rst?plain=1#L69)

> According to https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/bufferreg.html, there is a zero-copy speed benefit when the application uses ncclMemAlloc'ed buffers. Is "buffer registration cache" made for that?

I think this case is different. It is related to switch-based reduction, and ncclMemAlloc guarantees some memory alignment on the buffers (and probably also register them).

> Which "backend" exactly does the "caching", ProcessGroup or NCCL?

NCCL or UCC, or even lower level. There is no such cache at the PG level that I'm aware of.

Reply:
I'm not as an expert as you about RDMA and NIC. What's the implication of "buffer registration cache" for nvFuser or any DL compiler that tries to efficiently transfer data between multiple GPUs? 

I believe we already agree that using ncclMemAlloc'ed buffers or using slices within a larger ncclMemAlloc'ed region with proper alignment will speed things up. nvFuser has done this yet because it uses `at::empty`, but I believe this has to be solved. 

In addition to that, do you see extra benefit (and if so how much) of reusing the same base address and size iteation-to-iteration for an nccl call like allgather? The name "buffer registration cache" seems to say yes, but I'm not sure.

Reply:
FYI, @Aidyn-A pointed me to a [proposal](https://github.com/pytorch/pytorch/issues/124807) that he and his teammates are working on, which I think will help solve the problem without too many changes to nvFuser's allocation mechanism. But let's revisit this topic once more progress is made there.